File: python/剑指29顺时针打印矩阵.py

# nxn矩阵
def spiralOrder(matrix: list) -> list:
    res=[]
    n=len(matrix)
    
    for i in range(0,n>>1):
        
        for j in range(i,n-i-1):
            res.append(matrix[i][j])
        for j in range(i,n-i-1):
            res.append(matrix[j][n-i-1])
        for j in range(n-i-1,i,-1):
            res.append(matrix[n-i-1][j])
        for j in range(n-i-1,i,-1):
            res.append(matrix[j][i])

    if(n%2==1):
        res.append(matrix[n//2][n//2])
    return res

print(spiralOrder([[1,2,3],[4,5,6],[7,8,9]]))
'''
1   2   3   4
5   6   7   8
9   10  11  12
13  14  15  16 
'''

File: csharp/剑指29顺时针打印矩阵.cs
public partial class Solution
{
    public int[] SpiralOrder(int[][] matrix)
    {
        if (matrix.Length == 0)
        {
            return new int[0];
        }
        int[] res = new int[matrix[0].Length * matrix.Length];
        int l = 0, b = matrix.Length - 1, r = matrix[0].Length - 1, t = 0;
        int num = matrix[0].Length * matrix.Length;
        int k = 0;
        while (k < num)
        {
            for (int i = l; i <= r & k < num; i++)
                res[k++] = matrix[t][i];
            t++;
            
            for (int i = t; i <= b & k < num; i++)
                res[k++] = (matrix[i][r]);
            r--;
            
            for (int i = r; i >= l & k < num; i--)
                res[k++] = (matrix[b][i]);
            b--;
            
            for (int i = b; i >= t & k < num; i--)
                res[k++] = (matrix[i][l]); 
            l++;
        }
        return res;
    }

}